### ISSUE TITLE
How to retrieve an APIKey object based on a key?

### ISSUE THREAD

**Issue by jeancochrane** [florimondmanca/djangorestframework-api-key#92]
What's the recommended way of retrieving an `APIKey` instance based on a user-provided key? My use case is that I'd like to track the number of requests sent by each API key in my application.

The `BaseAPIKeyManager` does a similar lookup in its `is_valid()` method, where it seems like it's querying for the `APIKey` based on its `prefix` and then using the object instance's `is_valid()` method to double-check the key against the hashed version:

https://github.com/florimondmanca/djangorestframework-api-key/blob/851aec1fa88dc2f8a5594bf3c86a2fbd9c1588f6/src/rest_framework_api_key/models.py#L43-L59

Is it safe to perform a similar operation in my application (i.e. lookup the `APIKey` based on the prefix of the key provided by the user, and then double-check it using the `APIKey.is_valid()` method)?

Related to #20.

**Comment by florimondmanca** [comment]
Hi!

Yes, I assume getting the API key by prefix the same way is the safest way. You don’t want to be manipulating expired or invalid keys, so it’s good to always double check. (Also, the reason we query by prefix and not by the whole key is because only the prefix is stored in the DB, while the key itself is stored in an encrypted form.)

> Related to #92.

I don’t think you were meaning to recursively link to this issue? :)

Closing this but feel free to continue discussion.

**Comment by jeancochrane** [comment]
Thanks for the speedy response, I appreciate it!

> Yes, I assume getting the API key by prefix the same way is the safest way.

Great, that sounds good. Here's how I wound up implementing the lookup on my custom manager:

```python
    def get_from_key(self, key):
        prefix, _, _ = key.partition(".")
        queryset = self.get_usable_keys()
        api_key = queryset.get(prefix=prefix)
        if not api_key.is_valid(key):
            raise ValidationError('API key is not valid.')
        else:
            return api_key
```

Would you be interested in a PR that adds this method to `BaseAPIKeyManager`, refactors `BaseAPIKeyManager.is_valid()` to share the logic, and adds tests/docs for the new method? It seems like an operation that might be common enough to be a useful part of the core library.

> I don’t think you were meaning to recursively link to this issue? :)

Oops, yes -- fixed!

**Comment by florimondmanca** [comment]
> Would you be interested in a PR

Sure :-) Let's see how that looks like!

**Comment by florimondmanca** [comment]
Reopening since we're treating this as an active issue now, that could be fixed by #93. :-)

**Comment by cristianrat** [comment]
Sorry to reopen this (should I create a new issue?)
Using this with a bespoke model (ie: OrganisationAPIKey, so each of my keys belongs to an Organisation) throws
`django.core.exceptions.FieldError: Cannot resolve keyword 'revoked' into field. Choices are:`
That field is there, though
Am I using this wrong?
